File: test_src/fortran/hierarchical_parallelism/reduction/target_teams_distribute_parallel_for.f90

program target_teams_distribute_parallel_for
    USE OMP_LIB

    implicit none
  
    INTEGER :: L = 5
    INTEGER :: i
    
    REAL :: COUNTER = 0

    
    
     
    
    !$OMP TARGET TEAMS DISTRIBUTE PARALLEL DO   REDUCTION(+:COUNTER)   MAP(TOFROM: COUNTER) 


    DO i = 1 , L 


    

    


counter = counter + 1.

 
     

    END DO

    !$OMP END TARGET TEAMS DISTRIBUTE PARALLEL DO
    

    IF  ( ( ABS(COUNTER - L) ) > 10*EPSILON(COUNTER) ) THEN
        write(*,*)  'Expected L Got', COUNTER
        call exit(1)
    ENDIF

end program target_teams_distribute_parallel_for

File: test_src/cpp/hierarchical_parallelism/reduction/complex_float/target__teams_distribute_parallel_for.cpp
#include <iostream>
#include <cmath>
#include <stdexcept>

#include <complex>
using namespace std;




bool almost_equal(complex<float> x, complex<float> gold, float tol) {
    
        return abs(gold) * (1-tol) <= abs(x) && abs(x) <= abs(gold) * (1 + tol ); 
    
}


#pragma omp declare reduction(+: complex<float>:   omp_out += omp_in) 


void test_target__teams_distribute_parallel_for(){

 // Input and Outputs
 
 const int L = 5;

complex<float> counter{};

// Main program

#pragma omp target   map(tofrom:counter) 

{


#pragma omp teams distribute parallel for  reduction(+: counter)  

    for (int i = 0 ; i < L ; i++ )

{




counter += complex<float> { 1.0f };



}

}


// Validation
if ( !almost_equal(counter,complex<float> { L }, 0.1)  ) {
    std::cerr << "Expected: " << L << " Got: " << counter << std::endl;
    throw std::runtime_error( "target__teams_distribute_parallel_for give incorect value when offloaded");
}

}
int main()
{
    test_target__teams_distribute_parallel_for();
}
